README.TXT
生成可执行应用程序（绿色版免安装）
1、获得site-packages\PyQt5\Qt\bin目录
2、进入C:\Windows\System32，右键cmd.exe管理员运行
3、cd 定位到工程目录下
4、pyinstaller -Dw --paths "C:\Users\movin\AppData\Local\Programs\Python\Python37\Lib\site-packages\PyQt5\Qt\bin" log_analysis_main.spec

CHANGELOG
### [1.11_Beta] - Oct 29, 2020
**优化**
- 将之前记录的偏移量迁移到到config_offset文件，软件可自动识别不同版本的地址偏移量，当有新增偏移量时，可直接修改config_offset文件

**修复**
- 修改多用户冲突情况的，真实阻抗的识别字段
- 删除帮助文件地址偏移量，迁移到config_offset

### [1.10_Beta] - Oct 12, 2020
**新增**
- 新增v63版本以上Excel解析方法，增加字段[平滑阻抗, 平滑体脂率, 骨骼肌]，更新title
- v63蓝牙版本偏移量识别，偏移量为:0x1FFEE850

### [1.9_Beta] - Sep 28, 2020
**新增**
- flow log日志中只要出现体重，如果没有阻抗或者测量信息，也作为一条数据记录到excel -> v36InfoExtraction

**优化**
- 检索flow log的唯一关键字

### [1.8_Beta] - Sep 25, 2020
**新增**
- v51蓝牙版本偏移量识别，偏移量为:0x1FFEE890
- v53蓝牙版本偏移量识别，偏移量为:0x1FFEE880
- v53版本以上在多用户冲突下，流水日志新增了res打点，对日志中已打点的res值添加进Excel表

### [1.7_Beta] - Sep 21, 2020
**修复**
- Excel表部分数据没有日期时间，修复为即使没有用户信息也会打印日期时间
- 地址所对应的Log字符串出现6个值，但流水日志只保存4个参数值，导致字符串格式化错误，未打出对应Log字符串

### [1.6_Beta] - Sep 21, 2020
**新增**
- 新增菜单栏关于、版本更新日志、帮助
- 新增针对于v36版本以上TXT、Excel解析方法 -> v36InfoExtraction
- 导入BLE BIN文件时，自动识别BLE的版本号和已知偏移量，未识别到的版本，需自行选择版本信息和输入正确偏移量

**修复**
- BLE OTA版本为1.0.1.31之后，在用户识别标识行中会打印出2个值，后一个才为用户识别标识，更改逻辑判断
- 解析的TXT日志中，地址所对应的Log字符串格式化出现报错，将未格式化的字符串写入TXT
***

### [1.5_Beta] - Sep 12, 2020  
**优化**
- 优化检测到未识别信息时，Excel不对本条信息进行展示
***

### [1.4_Beta] - Sep 10, 2020
**新增**
- Excel新增用户识别（已识别、未识别、冲突识别）
- Excel新增一张Sheet表，录入ADC值

**优化**
- 优化Excel显示格式
***

### [1.3_Beta] - Sep 8, 2020
**新增**
- 新增选择文件时，默认打开用户上次选择路径
- 新增打开文件夹按键，可打开Excel和TXT所在的文件夹
- 新增解锁按键，防止误更改地址偏移量
- 新增恢复默认功能，可初始化地址偏移量

**优化**
- 优化TXT文件显示格式
- 优化online body fat未打点时的状态识别

**修复**
- 修复解析WIFI数据异常
***

### [1.2_Beta] - Sep 3, 2020  
**新增**
- 新增可选择更改地址偏移量  
***

### [1.1_Beta] - Sep 2, 2020  
**新增**
- 实现Excel和TXT文件同步输出
***

### [1.0_Beta] - Sep 1, 2020  
**新增**
- 初版  

flow_log_analysis.py
from constant import *
import time

# FLOW LOG LEVEL ANALYSIS
def level(level_data):
    if level_data == LOG_FLOW_LEVEL_FATAL:
        string = 'Fatal'
    elif level_data == LOG_FLOW_LEVEL_ERROR:
        string = 'Error'
    elif level_data == LOG_FLOW_LEVEL_WARN:
        string = 'Warn'
    elif level_data == LOG_FLOW_LEVEL_INFO:
        string = 'Info'
    elif level_data == LOG_FLOW_LEVEL_DEBUG:
        string = 'Debug'
    else:
        string = 'Flow log level error'
    return string

# FLOW LOG MODE ANALYSIS
def mode(mode_data):
    if 0 <= mode_data <= 15:
        string = MODE_BLE
    elif 16 <= mode_data <= 31:
        string = MODE_WIFI
    else:
        string = 'Flow log mode error'
    return string

# FLOW LOG TIME ANALYSIS
def time_(timeStamp_data):
    timeArray = time.localtime(timeStamp_data)
    date = time.strftime("%Y-%m-%d", timeArray)
    time_ = time.strftime("%H:%M:%S", timeArray)
    return date, time_

# FLOW LOG ADDRESS ANALYSIS
def addr(address_data, addr_offset, f_bin):
    address = address_data - addr_offset
    if address < 0:
        string = 'Flow log address error'
    elif address >= len(f_bin):
        string = 'Flow log address error'
    else:
        try:
            buf = {}
            s = 0
            while (f_bin[address] != 0x00):
                if f_bin[address] != 0x0a and f_bin[address] != 0x0d:
                    buf[s] = f_bin[address]
                address += 1
                s += 1
            buf_temp1 = bytes(buf.values())
            string = buf_temp1.decode()
        except Exception as e:
            string = 'Analysis error -> 0x{:02X}'.format(address_data)
    return string

# FLOW LOG PARA_1 ANALYSIS
def fileIndexAndLineNum(para1_data):
    index = para1_data >> 16
    lineNum = para1_data & 0x0000FFFF
    string = ('%d' % index), ('%d' % lineNum)
    return string

# FLOW LOG PARA_2/3/4/5 ANALYSIS
def para2_5(para2_data, para3_data, para4_data, para5_data):
    string = ('%d' % para2_data), ('%d' % para3_data), ('%d' % para4_data), ('%d' % para5_data)
    return string

log_analysis_main.py
import os
import time
import traceback

from PyQt5.Qt import *
import sys
from UI.main_ui import Ui_main
from log_analysis_tool import logFileManage
from logger import Logger

# app information
APP_NAME = 'FLOW LOG Analysis'
APP_VERSION = '1.11_Beta'
BUILT_DATE = 'Built on Oct 29, 2020'
POWERED_BY = 'Powered by: Vincoo'
EMAIL = 'Email: movincoo@163.com | matengnan@ebelter.com'

MREGE_TEMP_FILE_NAME = './mrege_temp_file'
FONT_LABEL = '微软雅黑'
DEFAULT_BLE_ADDROFFSET = '0x1FFEED80'
DEFAULT_WIFI_ADDROFFSET = '0x40210000'
CONFIG_OFFSET_PATH = "./config_offset.ini"

LOG = Logger(level=Logger.DEBUG).get_logger()


class MainWindow(QMainWindow, Ui_main):
    def __init__(self, parent=None, *args, **kwargs):
        super(MainWindow, self).__init__(parent, *args, **kwargs)
        self.setupUi(self)
        self.setWindowFlags(Qt.MSWindowsFixedSizeDialogHint)
        self.setMaximumSize(self.width(), self.height())
        self.setMinimumSize(self.width(), self.height())
        self.setWindowTitle(APP_NAME + '_' + APP_VERSION)
        self.logpath = ''
        self.config_setting = None
        self.initView()
        self.bleVersion = 0
        # 测试使用

        # logFilePath = r'C:/Users/Vincoo/Desktop/HUAWEI BH30 流水日志解析项目/Log/Log20200902/Herm-B19_1.0.1.24_BF844BD274CB13D64BC0B621215C9A028414056552A88672C3A51ED32BB1486B_20200902155916_WearableBeta_mcu_debug.log'
        # self.logFilePath = r'C:/Users/Vincoo/Desktop/HUAWEI BH30 流水日志解析项目/Log/Log20200902/Herm-B19_1.0.1.24_126A7EED3B4DAD503125B3EECFF9747FBA15B6BC8FC66584D952DFF95530744D_20200902235919_WearableBeta_mcu_debug.log'
        # self.bleBinFilePath = r'C:/Users/Vincoo/Desktop/HUAWEI BH30 流水日志解析项目/Log/Log20200902/Herm_BLE_V1.0.1.24.bin'
        # self.wifiBinFilePath = r'C:/Users/Vincoo/Desktop/HUAWEI BH30 流水日志解析项目/Log/Log20200902/Herm_WIFI_V1.0.1.24A_COMBINE.bin'
        # self.bleAddrOffset = 0x1FFEED80
        # self.wifiAddrOffset = 0x40210000

    def initView(self):
        # addr offset恢复默认
        self.addrOffsetDefault()
        self.bt_select_logfile.clicked.connect(self.selectLogFiles)
        self.bt_add_ble_binfile.clicked.connect(self.selectBleBinFile)
        self.bt_add_wifi_binfile.clicked.connect(self.selectWifiBinFile)
        self.bt_analysis.clicked.connect(self.analysisFile)

        self.bt_lock.clicked.connect(self.addrOffsetChangeLock)
        self.bt_default.clicked.connect(self.addrOffsetDefault)
        self.bt_opendir.clicked.connect(self.openDir)

        self.actionAbout.triggered.connect(self.showAboutDialog)
        self.actionUpdateInfo.triggered.connect(self.showChangeLogDialog)
        self.actionHelp.triggered.connect(self.showHelpDialog)

        self.rb_36down.setHidden(True)
        self.rb_36up.setHidden(True)

    def showAboutDialog(self):
        '''
            菜单栏显示关于窗口
        '''
        self.aboutDialog = QDialog(self, flags=Qt.WindowCloseButtonHint | Qt.MSWindowsFixedSizeDialogHint)
        self.aboutDialog.resize(480, 180)
        self.aboutDialog.setWindowTitle('关于')

        self.aboutLabel_0 = QLabel()
        self.aboutLabel_0.setText(APP_NAME)
        self.aboutLabel_0.setFont(QFont(FONT_LABEL, 10, QFont.Bold))

        self.aboutLabel_1 = QLabel()
        self.aboutLabel_1.setText('Version: ' + APP_VERSION)
        self.aboutLabel_1.setFont(QFont(FONT_LABEL, 9))

        self.aboutLabel_2 = QLabel()
        self.aboutLabel_2.setText(BUILT_DATE)
        self.aboutLabel_2.setFont(QFont(FONT_LABEL, 9))

        self.aboutLabel_3 = QLabel()
        self.aboutLabel_3.setText(POWERED_BY)
        self.aboutLabel_3.setFont(QFont(FONT_LABEL, 9))

        self.aboutLabel_4 = QLabel()
        self.aboutLabel_4.setText(EMAIL)
        self.aboutLabel_4.setFont(QFont(FONT_LABEL, 9))

        self.vbox = QVBoxLayout()
        self.vbox.addStretch()
        self.vbox.addWidget(self.aboutLabel_0)
        self.vbox.addWidget(self.aboutLabel_1)
        self.vbox.addWidget(self.aboutLabel_2)
        self.vbox.addWidget(self.aboutLabel_3)
        self.vbox.addWidget(self.aboutLabel_4)
        self.vbox.addStretch()
        self.vbox.setContentsMargins(30, 0, 30, 0)
        self.aboutDialog.setLayout(self.vbox)
        self.aboutDialog.open()  # open窗口模态，只会阻塞一个窗口，而不是将整个系统阻塞掉

    def showChangeLogDialog(self):
        '''
            菜单栏显示软件版本更新日志
        '''
        self.changeLogDialog = QDialog(self, flags=Qt.WindowCloseButtonHint | Qt.MSWindowsFixedSizeDialogHint)
        self.changeLogDialog.resize(480, 420)
        self.changeLogDialog.setWindowTitle('版本更新日志')
        self.changeLog_text = QTextBrowser(self.changeLogDialog)
        self.changeLog_text.resize(480, 420)
        self.changeLog_text.setFont(QFont(FONT_LABEL, 10))
        self.changeLog_text.setMarkdown(open('./CHANGELOG', mode='r', encoding='utf8').read())
        # self.changeLog_text.setText()
        self.changeLog_text.setStyleSheet('background-color: rgb(240, 240, 240);')  # rgb(240, 240, 240)
        self.changeLog_text.setFrameShape((QFrame.NoFrame))

        self.changeLogDialog.open()

    def showHelpDialog(self):
        '''
           菜单栏显示帮助
        '''
        self.helpDialog = QDialog(self, flags=Qt.WindowCloseButtonHint | Qt.MSWindowsFixedSizeDialogHint)
        self.helpDialog.resize(480, 420)
        self.helpDialog.setWindowTitle('帮助')
        self.help_text = QTextBrowser(self.helpDialog)
        self.help_text.resize(480, 420)
        self.help_text.setFont(QFont(FONT_LABEL, 10))
        self.help_text.setMarkdown(open('./HELP', mode='r', encoding='utf8').read())
        # self.changeLog_text.setText()
        self.help_text.setStyleSheet('background-color: rgb(240, 240, 240);')  # rgb(240, 240, 240)
        self.help_text.setFrameShape((QFrame.NoFrame))

        self.helpDialog.open()

    def addrOffsetDefault(self):
        self.le_ble_addr_offset.setText(DEFAULT_BLE_ADDROFFSET)
        self.le_wifi_addr_offset.setText(DEFAULT_WIFI_ADDROFFSET)

    def addrOffsetChangeLock(self):
        if self.bt_lock.text() == '解锁':
            self.le_ble_addr_offset.setEnabled(True)
            self.le_wifi_addr_offset.setEnabled(True)
            self.bt_lock.setText('加锁')
        else:
            self.le_ble_addr_offset.setEnabled(False)
            self.le_wifi_addr_offset.setEnabled(False)
            self.bt_lock.setText('解锁')

    def openDir(self):
        if os.path.exists('./OutFile'):
            os.startfile('.\\OutFile')
        else:
            QMessageBox.warning(self, '错误', '未找到文件夹路径，打开文件夹失败！', QMessageBox.Ok)

    def selectLogFiles(self):
        lastPath = self.getConfigInfo()
        path = QFileDialog.getOpenFileNames(self, '选择Log文件', lastPath, 'ALL(*.*);;log files(*.txt *.log)',
                                            'log files(*.txt *.log)')[0]

        if path != []:
            self.setConfigInfo(path[0])
            self.logpath = path
            self.list_file_path.clear()
            try:
                for i, val in enumerate(self.logpath):
                    self.list_file_path.addItem(str(i + 1) + ' | ' + val)
            except Exception as e:
                LOG.error(repr(e))

    def selectBleBinFile(self):
        lastPath = self.getConfigInfo()
        path = QFileDialog.getOpenFileName(self, '选择蓝牙烧录文件', lastPath, 'ALL(*.*);;firmware files(*.bin)',
                                           'firmware files(*.bin)')[0]
        if path != '':
            self.setConfigInfo(path)
            self.le_ble_binfile.setText(path)
            blefilename = os.path.basename(path)
            if 'BLE' in blefilename:
                try:
                    ble_version = blefilename.rstrip('.bin').split('V')[1]
                    self.bleVersion = int(ble_version.split('.')[3], 10)
                except Exception as e:
                    self.lb_version.setText('[未识别，请选择版本信息！]')
                    self.rb_36down.setHidden(False)
                    self.rb_36up.setHidden(False)
                    QMessageBox.information(self, '提示', '版本号未能自动识别，请选择BIN版本信息，确认蓝牙地址偏移量是否正确！', QMessageBox.Ok)
                else:
                    self.lb_version.setText(
                        '[已识别版本] -%s- [已设置偏移量为%s]' % (ble_version, self.getBleAddrOffset(self.bleVersion)))
                    self.rb_36down.setHidden(True)
                    self.rb_36up.setHidden(True)

            else:
                self.lb_version.setText('[未识别，请选择版本信息！]')
                self.rb_36down.setHidden(False)
                self.rb_36up.setHidden(False)
                QMessageBox.information(self, '提示', '版本号未能自动识别，请选择BIN版本信息，确认蓝牙地址偏移量是否正确！', QMessageBox.Ok)

    def selectWifiBinFile(self):
        lastPath = self.getConfigInfo()
        path = QFileDialog.getOpenFileName(self, '选择WIFI烧录文件', lastPath, 'ALL(*.*);;firmware files(*.bin)',
                                           'firmware files(*.bin)')[0]
        if path != '':
            self.setConfigInfo(path)
            self.le_wifi_binfile.setText(path)

    def getConfigInfo(self):
        self.config_setting = QSettings("./config_setting.ini", QSettings.IniFormat)
        path = str(self.config_setting.value('LastFilePath'))
        if path is None or path == '':
            return './'
        return path

    def setConfigInfo(self, path):
        self.config_setting.setValue('LastFilePath', path)

    def setOffsetConfigInfo(self):
        LOG.debug('创建默认Config')
        self.config_offset.setValue('0', '0x1FFEED80')
        self.config_offset.setValue('28', '0x1FFEED90')
        self.config_offset.setValue('31', '0x1FFEE920')
        self.config_offset.setValue('51', '0x1FFEE890')
        self.config_offset.setValue('53', '0x1FFEE880')
        self.config_offset.setValue('63', '0x1FFEE850')
        self.config_offset.setValue('81', '0x1FFEE8A0')

    def getBleAddrOffset(self, version):
        offset = ''
        self.config_offset = QSettings(CONFIG_OFFSET_PATH, QSettings.IniFormat)
        key_list = list(map(int, self.config_offset.allKeys()))
        if len(key_list) == 0:
            # 创建默认config
            self.setOffsetConfigInfo()
            key_list = list(map(int, self.config_offset.allKeys()))
        key_list.sort()
        LOG.debug('列表sort之后的key值%s' % key_list)
        if version in key_list:
            offset = self.config_offset.value(str(version))
        else:
            max_key = max(key_list)
            if max_key < version:
                offset = self.config_offset.value(str(max_key))
            else:
                for index, config_key in enumerate(key_list):
                    if config_key > version:
                        offset = self.config_offset.value(str(key_list[index - 1]))
                        break
        self.le_ble_addr_offset.setText(offset)
        return offset

    def analysisFile(self):
        ble_addr_offset = self.le_ble_addr_offset.text()
        wifi_addr_offset = self.le_wifi_addr_offset.text()
        if not all([ble_addr_offset, wifi_addr_offset]):
            QMessageBox.warning(self, '错误', '偏移量输入错误', QMessageBox.Ok)
            return
        bin_ble_path = self.le_ble_binfile.text()
        bin_wifi_path = self.le_wifi_binfile.text()
        if not all([self.logpath, bin_ble_path, bin_wifi_path]):
            QMessageBox.warning(self, '错误', '未选择文件！', QMessageBox.Ok)
            return
        if self.lb_version.text() == '[未识别，请选择版本信息！]':
            if not any([self.rb_36up.isChecked(), self.rb_36down.isChecked()]):
                QMessageBox.warning(self, '错误', '版本号未能自动识别，请选择BIN版本！', QMessageBox.Ok)
                return
            else:
                if self.rb_36down.isChecked():
                    self.bleVersion = 0
                if self.rb_36up.isChecked():
                    self.bleVersion = 36
        try:
            if len(self.logpath) == 1:
                log_path = self.logpath[0]
            else:
                if os.path.exists(MREGE_TEMP_FILE_NAME):
                    os.remove(MREGE_TEMP_FILE_NAME)
                tempfile = open(MREGE_TEMP_FILE_NAME, 'ab+')
                for path in self.logpath:
                    f = open(path, 'rb')
                    s = f.read()
                    tempfile.write(s)
                    tempfile.flush()
                    f.close()
                tempfile.close()
                log_path = MREGE_TEMP_FILE_NAME
        except Exception as e:
            LOG.error(repr(e))
        else:
            # 开始解析
            try:
                self.analysis_progressBar_dialog()
                self.analysisThread = AnalysisThread(log_path, bin_ble_path, bin_wifi_path, int(ble_addr_offset, 16),
                                                     int(wifi_addr_offset, 16), self.bleVersion)
                self.analysisThread.total_signal.connect(self.total_callBack)
                self.analysisThread.analysis_signal.connect(self.analysis_callBack)
                self.analysisThread.start()
            except Exception as e:
                time.sleep(0.5)
                self.progressBarDialog.close()
                QMessageBox.warning(self, '提示', '失败，解析出错，请将对应日志和BIN文件发送开发者！', QMessageBox.Ok)
                LOG.error(e)

    def total_callBack(self, number):
        self.progressBar.setRange(0, number)

    def analysis_callBack(self, number):
        if number == -1:
            time.sleep(0.5)
            self.progressBarDialog.close()
            QMessageBox.information(self, '提示', '解析结束', QMessageBox.Ok)
        elif number == -2:
            time.sleep(0.5)
            self.progressBarDialog.close()
            QMessageBox.warning(self, '提示', '失败，解析出错，请将对应日志和BIN文件发送至开发者！', QMessageBox.Ok)
        else:
            self.progressBar.setValue(number + 1)

    def analysis_progressBar_dialog(self):
        # Dialog
        self.progressBarDialog = QDialog(self, flags=Qt.FramelessWindowHint | Qt.MSWindowsFixedSizeDialogHint)
        self.progressBarDialog.resize(400, 150)
        # Label
        self.lb_state = QLabel(self.progressBarDialog)
        self.lb_state.setContentsMargins(0, 0, 0, 10)
        self.lb_state.setFont(QFont(FONT_LABEL, 9))
        # ProgressBar
        self.progressBar = QProgressBar(self.progressBarDialog)
        # self.progressBar.setRange(0, 100)
        # Layout
        self.vbox_download = QVBoxLayout(self.progressBarDialog)
        self.vbox_download.addStretch()
        self.vbox_download.addWidget(self.lb_state)
        self.vbox_download.addWidget(self.progressBar)
        self.vbox_download.addStretch()
        self.vbox_download.setContentsMargins(30, 0, 30, 15)

        self.progressBarDialog.setLayout(self.vbox_download)
        self.progressBarDialog.open()
        self.lb_state.setText('正在解析，请稍后……')


class AnalysisThread(QThread):
    analysis_signal = pyqtSignal(int)
    total_signal = pyqtSignal(int)

    def __init__(self, logFilePath, bleBinPath, wifiBinPath, bleAddrOffset, wifiAddrOffset, binVersion):
        super(AnalysisThread, self).__init__()
        self.logFilePath = logFilePath
        self.bleBinPath = bleBinPath
        self.wifiBinPath = wifiBinPath
        self.bleAddrOffset = bleAddrOffset
        self.wifiAddrOffset = wifiAddrOffset
        self.binVersion = binVersion

    def run(self):
        try:
            logFileManage(self.logFilePath, self.bleBinPath, self.wifiBinPath, self.bleAddrOffset, self.wifiAddrOffset,
                          self.analysis_signal, self.total_signal, self.binVersion)
            self.analysis_signal.emit(-1)
        except Exception as e:
            self.analysis_signal.emit(-2)
            traceback.print_exc()


if __name__ == '__main__':
    app = QApplication(sys.argv)
    mainWin = MainWindow()
    mainWin.show()
    LOG.info('Version: %s' % APP_VERSION)
    LOG.info('Time: %s' % time.strftime('%Y-%m-%d %H:%M:%S', time.localtime()))
    sys.exit(app.exec_())
